plugin-core/src/main/java/io/singularitynet/sdk/plugin/ServiceApiGetter.java

```diff
@@ -0,0 +1,205 @@
+                registryContract, metadataStorage);
+        ServiceMetadata metadata = metadataProvider.getServiceMetadata();
+        log.debug("service metadata: {}", metadata);
+        loadAndUnpackApi(metadataStorage, metadata.getModelIpfsHash());
+    }
+
+    private Registry getRegistryContract(Web3j web3j) throws PluginException {
+        String networkId;
+        try {
+            networkId = web3j.netVersion().send().getNetVersion();
+        } catch (IOException e) {
+            throw new PluginException("Could not get Ethereum network id", e);
+        }
+        DefaultGasProvider gasProvider = new DefaultGasProvider();
+        ReadonlyTransactionManager transactionManager = new ReadonlyTransactionManager(
+                web3j, new Address(parameters.getGetterEthereumAddress()).toString());
+        String registryAddress = parameters.getRegistryAddress();
+        if (registryAddress == null || registryAddress.isEmpty()) {
+            registryAddress = ContractUtils.readContractAddress(networkId, "Registry").toString();
+        }
+        Registry registry = Registry.load(new Address(registryAddress).toString(),
+                web3j, transactionManager, gasProvider);
+        return registry;
+    }
+
+    private void loadAndUnpackApi(IpfsMetadataStorage metadataStorage, String ipfsHash) throws PluginException {
+        URI uri;
+        try {
+            uri = new URI("ipfs://" + ipfsHash);
+            log.info("IPFS uri: {}", uri);
+        } catch (URISyntaxException e) {
+            throw new PluginException("Incorrect IPFS hash in metadata: " + ipfsHash, e);
+        }
+        byte[] apiTar = metadataStorage.get(uri);
+        log.info("API size {} bytes", apiTar.length);
+        unpackApi(apiTar);
+    }
+
+    private void unpackApi(byte[] apiTar) throws PluginException {
+        ArchiveInputStream is = new TarArchiveInputStream(
+                new ByteArrayInputStream(apiTar));
+        try {
+            ArchiveEntry entry = null;
+            while ((entry = is.getNextEntry()) != null) {
+                if (!is.canReadEntryData(entry)) {
+                    throw new PluginException("Cannot read entry in API archive: " + entry);
+                }
+                log.info("Unpacking {}", entry.getName());
+                String name = new File(parameters.getOutputDir(), entry.getName()).getAbsolutePath();
+                File f = new File(name);
+                if (entry.isDirectory()) {
+                    if (!f.isDirectory() && !f.mkdirs()) {
+                        throw new IOException("failed to create directory " + f);
+                    }
+                } else {
+                    File parent = f.getParentFile();
+                    if (!parent.isDirectory() && !parent.mkdirs()) {
+                        throw new IOException("failed to create directory " + parent);
+                    }
+                    try (OutputStream o = Files.newOutputStream(f.toPath())) {
```

Comment:
Does a **try** without **catch** does anything?

Reply:
Yes, it is `try with` block, it closes `OutputStream o` after leaving the block. So it is equal to 
```java
OutputStream o = Files.newOutputStream(f.toPath());
try {
} finally {
  o.close();
}
```